ir/src/constraints/graph.rs

```diff
@@ -398,6 +405,11 @@ impl AlgebraicGraph {
                 Ok(ExprDetails::new(node_index, trace_segment, domain))
             }
             IdentifierType::PublicInput(_) => {
+                if !domain.is_boundary() {
```

Comment:
now that `is_boundary` is defined, it can be updated in `ConstrainedBoundary` as well. I think it makes sense to do this at the same time as defining the method, unless you object